LazyOptional no longer silently swallows all Throwables

Gamebuster19901 opened the pull request:
LazyOptional currently uses try-catch block for control flow. It also catches Throwable and swallows it silently. This is bad practice. This pull request fixes the issue by using an if-statement instead of a try-catch block for control flow. This change may also improve the performance of LazyOptional slightly, as we are no longer using a try-catch block.

This issue caused me not to be able to figure out why an Assertion I made was false on the dedicated server, but not on the client or integrated server.

As you can see from [my code](https://pastebin.com/tzw3asPs), it appears that it should be impossible for the AssertionError to occur.

Stacktrace from my code:
```
[04Jun2019 15:12:37.639] [Server thread/FATAL] [net.minecraft.server.MinecraftServer/]: Error executing task
java.util.concurrent.ExecutionException: java.lang.AssertionError
    at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:1.8.0_144]
    at java.util.concurrent.FutureTask.get(FutureTask.java:192) ~[?:1.8.0_144]
    at net.minecraft.util.Util.runTask(Util.java:127) [?:?]
    at net.minecraft.server.MinecraftServer.updateTimeLightAndEntities(MinecraftServer.java:770) [?:?]
    at net.minecraft.server.dedicated.DedicatedServer.updateTimeLightAndEntities(DedicatedServer.java:352) [?:?]
    at net.minecraft.server.MinecraftServer.tick(MinecraftServer.java:724) [?:?]
    at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:611) [?:?]
    at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
Caused by: java.lang.AssertionError
    at net.minecraftforge.common.util.LazyOptional.orElseThrow(LazyOptional.java:260) ~[forge-1.13.2-25.0.110_mapped_snapshot_20190409-1.13.2-recomp.jar:?]
    at com.gamebuster19901.halo.common.item.weapons.assaultrifle.AssaultRifle$1.<init>(AssaultRifle.java:72) ~[main/:?]
    at com.gamebuster19901.halo.common.item.weapons.assaultrifle.AssaultRifle.initCapabilities(AssaultRifle.java:70) ~[main/:?]
    at net.minecraft.item.ItemStack.forgeInit(ItemStack.java:1042) ~[forge-1.13.2-25.0.110_mapped_snapshot_20190409-1.13.2-recomp.jar:?]
    at net.minecraft.item.ItemStack.<init>(ItemStack.java:107) ~[forge-1.13.2-25.0.110_mapped_snapshot_20190409-1.13.2-recomp.jar:?]
    at net.minecraft.item.ItemStack.<init>(ItemStack.java:100) ~[forge-1.13.2-25.0.110_mapped_snapshot_20190409-1.13.2-recomp.jar:?]
    at net.minecraft.command.arguments.ItemInput.createStack(ItemInput.java:36) ~[forge-1.13.2-25.0.110_mapped_snapshot_20190409-1.13.2-recomp.jar:?]
    at net.minecraft.command.impl.GiveCommand.giveItem(GiveCommand.java:38) ~[forge-1.13.2-25.0.110_mapped_snapshot_20190409-1.13.2-recomp.jar:?]
    at net.minecraft.command.impl.GiveCommand.lambda$register$1(GiveCommand.java:25) ~[forge-1.13.2-25.0.110_mapped_snapshot_20190409-1.13.2-recomp.jar:?]
    at com.mojang.brigadier.CommandDispatcher.execute(CommandDispatcher.java:132) ~[brigadier-1.0.14.jar:?]
    at net.minecraft.command.Commands.handleCommand(Commands.java:209) ~[?:?]
    at net.minecraft.network.NetHandlerPlayServer.handleSlashCommand(NetHandlerPlayServer.java:1027) ~[?:?]
    at net.minecraft.network.NetHandlerPlayServer.processChatMessage(NetHandlerPlayServer.java:1007) ~[?:?]
    at net.minecraft.network.play.client.CPacketChatMessage.processPacket(CPacketChatMessage.java:40) ~[?:?]
    at net.minecraft.network.play.client.CPacketChatMessage.processPacket(CPacketChatMessage.java:8) ~[?:?]
    at net.minecraft.network.PacketThreadUtil.lambda$checkThreadAndEnqueue$0(PacketThreadUtil.java:15) ~[?:?]
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_144]
    at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_144]
    at net.minecraft.util.Util.runTask(Util.java:126) ~[?:?]
    ... 5 more
```

What actually happened was I was calling client code on the dedicated server, but forge was swallowing the exception, causing LazyOptional.getValue() to return null, which then caused orElseThrow() to throw the provided AssertionError:

![image](https://user-images.githubusercontent.com/7155747/58927738-caef6a00-871d-11e9-91bd-8d13409c42b6.png)

Gamebuster19901 committed:
LazyOptional no longer swallows exceptions

Gamebuster19901 committed:
formatting

LexManos commented:
Whops, the intention wasnt to use it as a control flow, that was just a little misorganization of the code.
The intention WAS however to eat any non-state exceptions thrown by the supplier so that the consumer doesn't get random errors when the supplier screws up like yours does.

Perhaps a better solution may be to add a hasErrored/getError pair.

mezz commented:
In that case it may be best to always catch and log the error, nobody calling it should really be concerned with if it's crashing.

Gamebuster19901 committed:
Don't return null so the source of errors are clearer.

Gamebuster19901 committed:
"if it has no supplier or it is invalid, it should return null"

mezz commented on src/main/java/net/minecraftforge/common/util/LazyOptional.java:
```diff
@@ -29,6 +29,10 @@
 
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
```
I think this is no longer needed

Gamebuster19901 replied:
Whoops, I forgot to remove the logger. Done.

mezz commented on src/main/java/net/minecraftforge/common/util/LazyOptional.java:
```diff
@@ -295,4 +295,4 @@ public void invalidate()
         this.listeners.forEach(e -> e.accept(this));
     }
-}
+}
```
please undo this minor change

Gamebuster19901 replied:
Eclipse seems to want to keep removing the newline character, I had to add it manually with Notepad++.

mezz reviewed (changes requested):
mostly looks good to me, just some minor cleanup to do

Gamebuster19901 committed:
Remove unnecessary logger

LexManos commented on src/main/java/net/minecraftforge/common/util/LazyOptional.java:
```diff
@@ -108,18 +108,13 @@ private LazyOptional(@Nullable NonNullSupplier<T> instanceSupplier)
-            catch (Throwable e)
-            {
-                return null;
+                throw new NullPointerException("Supplier should not return null value");
```
We do not want to throw an exception to be caught by the consumer. Logging it noisily is probably the way to go.

Gamebuster19901 replied:
I've reverted back to the previous version that does that.

Gamebuster19901 committed:
This squahsed commit reverts the following commits:

c45d08904d97c942507ddf2d157b1ec87533168a
40ab76d7dae6e722c1763b5b16fb6825457e93fa
dd175663428afa779bc0be2e35175564e5659122

Gamebuster19901 committed:
formatting
